h2o-core/src/main/java/water/util/VecUtils.java

```diff
@@ -863,4 +865,53 @@ public void map(Chunk c, NewChunk nc) {
     }
   }
 
+  /**
+   * Randomly shuffle a Vec using Fisher Yates shuffle 
+   * https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle
+   */
+  public static class ShuffleVecTask extends MRTask<ShuffleVecTask> {
+    private final long _seed;
+    public ShuffleVecTask(final long seed) {
+      super();
+      _seed = seed;
+    }
+
+    @Override public void map(Chunk cs, Chunk ncs) {
```

Comment:
@tomasfryda this means we never cross the chunk boundary, this could be an issue, wdyt?

Reply:
Good catch! Indeed, this could cause some issues, e.g., it could make sorted features less important.

Some papers recommend just splitting the dataset in half and swapping those halves, I wonder if we could do something similar - something like subclass `Vec` to `PermutedVec` and override `chunkForChunkIdx` so it permutes local chunks (so we don't have to send chunks between nodes). Or do you know of some better way?

Reply:
I think the biggest issue with the current method will be when you have a big dataset and sample it to a low number. Then some chunks can end up having just a single row and no shuffling would be done.

So I think something will need to be changed.

One thing you could do is to randomly shuffle the (whole) chunks (across the cluster). Eg. send chunk 1 to node X, chunk 2 to node Y...

Then in your procedure you would pick a random chunk from a local node and get a random value from that chunk.

While moving data from node to node is expensive, the more expensive part would actually be scoring per-feature. And the chunk shuffle can be done on the background while feature > 1 is being scored (for the first feature you would have to do it first obviously).

Reply:
> Then some chunks can end up having just a single row and no shuffling would be done.

Thanks for pointing this out. Now I have something that should mitigate that issue - basically:

1. In `setupLocal`: permute the array of local chunk indices
2. In `map`: Get the chunkId of the current chunk and using the permuted array of local chunks get another local chunk

This should make the behavior you described possible only if the value of `n_samples` is set too low (to the number of chunks and at the same time there would have to be only one chunk per node) which on the other hand  shouldn't matter since for those situations (`n_samples < 1000`) I use randomly selected rows using the `deepSlice` that you suggested so the user would have to have >1000 nodes and select the same or lower number of samples.

I think this implementation would be sufficient, however, I am also looking into the "shuffle chunks across the nodes" as that would provide us with "statistically nicer" permutation but I am unsure if it would provide any significant benefit as PFI most likely will not be run that many times in a row on one dataset to notice any issues with the statistical properties of this kind of permutation.
 
For now, I have locally implemented a version where I permute the chunks across the nodes and I also permute the content of each chunk (but only within the chunks boundary). But it is important to note that the main point is to break the ties between the rows and the rows of the permuted column which both implementations do.

Another thing I can do is to use your suggestion but I haven't come up with an efficient solution that would ensure we would really end up with permutation as it would be closer to a bootstrap sample (this wouldn't be a problem if the user decides to sample the data).

The next steps can be:
1. use the current implementation (permute local chunks and data within the individual chunks)
2. permute chunks across all the nodes and permute data within the individual chunks
3. use hierarchical random sampling you proposed


Personally, I'd go with (1) as it is the fastest ( (2) seem to be ~30% slower on my test-case on local 4 node cluster) and I think it is good enough (provides enough random permutations if all cases that I could think of, I often seen PFI used several times to get better sense of how confident we can be but the number of runs was somewhere between 0 and 50).

What do you think, @michalkurka ?

Reply:
@tomasfryda thanks for the analysis

I think it makes sense to go for (1) - permute local chunks and data within the individual chunks